### ISSUE TITLE
CLP not working as expected

### ISSUE THREAD

**Issue by PhillipWinder** [parse-community/parse-dashboard#784]
Make sure these boxes are checked before submitting your issue -- thanks for reporting issues back to Parse Dashboard!

- [X] You're running version >=1.0.23 of Parse Dashboard.

- [X] You're running version >=2.3.2 of Parse Server.

- [X] You've searched through [existing issues](https://github.com/ParsePlatform/parse-dashboard/issues?utf8=%E2%9C%93&q=). Chances are that your issue has been reported or resolved before.

#### Steps to reproduce

I have a class XYZ. If I go into the security menu of this class and set public read to FALSE everything works as expected. I can now not read any data from this class. However if I click on "Role, User or Pointer...", add my user with which I am currently logged in and set read to TRUE for this user i still cannot retrieve the data from this class. Do I have to do something differently here?

My goal would be to make certain data only visible to certain user (further down by assigning them to roles).

**Comment by flovilmart** [comment]
Can you provide the content of your schema?

**Comment by natanrolnik** [comment]
Please see my comment in #3481. This is the expected result. Without the master key, a query must pass **both** CLP and ACL in order to access an object, not one of them.

**Comment by natanrolnik** [comment]
Please see my comment on this [issue](https://github.com/parse-community/parse-server/issues/3481). 

This is the expected result. Without the master key, a query must pass **both CLP and ACL” checks to access an object.

**Comment by PhillipWinder** [comment]
Thank you for your answers. However I am still a little bit stuck here. If I want different roles to be able to "read" different classes, how do I implement that? For example user X should have read permission for class XYZ but not ABC and user Y should have read permission for both classes.

**Comment by natanrolnik** [comment]
Wait, it's my bad. I thought you were using ACLs with your user. CLPs should indeed work as you are describing. Are you able to post here the Schema as @flovilmart requested?

**Comment by PhillipWinder** [comment]
Wow, you guys are fast, thanks so much :)

What exactly do you mean by schema? How the classes are structured, their columns, the column types and so on?

**Comment by natanrolnik** [comment]
In your Mongo database, there is a collection called `_Schema`. Can you post the content of its single document?

**Comment by PhillipWinder** [comment]
I just found out, that it only happens for a specific class. In all other cases it works fine. I will debug a little bit and get back to you :)

**Comment by PhillipWinder** [comment]
Ok, so for this specific class I am getting the following error:

2017-10-19T15:02:52.146Z - Permission denied for action find on class XYZ.
2017-10-19T15:02:52.145Z - Error generating response. ParseError {
  code: 119,
  message: 'Permission denied for action find on class XYZ.' }

I am however never using the find method on the class XYZ.

**Comment by PhillipWinder** [comment]
Here is a snippet of my cloud code to demonstrate where the error is happening:

the class in question is called "Booking"

________

let bookingQuery = new Parse.Query("Booking");
            
            bookingQuery.equalTo("golfCourse", golfCourseParseObject);
            bookingQuery.equalTo("isCanceled", false);            
            bookingQuery.include("user");      
            // Until here, everything get executed as expected - no Errors
   
            bookingQuery.each(booking => {             
                // This does not execute - throws the error message

**Comment by natanrolnik** [comment]
@PhillipWinder how are you saving this booking object? Who has the ACL on it? You should pass the session token of the user you want. For example:

```
bookingQuery.find({sessionToken: user.getSessionToken()}).then(function(results) {
//iterate over results.
});
```

**Comment by PhillipWinder** [comment]
A booking object is set on the frontend (the user selects some parameters (date, golfcourse and so on) and then booking.save() is called to create the object.

ACL on all booking objects (as of now) is public read write true.

You mean like this?

    bookingQuery.find({sessionToken: user.getSessionToken()}).then(function(results) {
        bookingQuery.each(booking => {
            //iterate
    });

**Comment by natanrolnik** [comment]
> ACL on all booking objects (as of now) is public read write true.

If ACLs are public write and read, then how come the finds come empty? Is the CLP also public write and read? 

I think you might be missing something or the query isn’t right. Try running the same query if master key and see the response. If it has objects, then it’s a ACL/CLP “problem” (likely not really a problem in how they work, but in how you are expecting them to work); if it doesn’t return objects even with the master key, then the query parameters might be wrong.

**Comment by PhillipWinder** [comment]
“If ACLs are public write and read, then how come the finds come empty?“ - this is exactly my question :D

If I set the CLP for the booking class to public read write true everything works fine. If I set it to public read write false also everything works as expected (nothing can be accessed). If I set it too public read write false AND use the masterkey also everything works as expected (everything can be accessed).

BUT if I set the CLP to public read write false AND myUser read write true I cant get the data - I get the above mentioned error and the find/each method return null - that is the strange thing thats happening. Other query methods (include, equalTo..) work fine (I can access the data with my user) but find/each strangely do not work.
